cpp/misra/src/rules/RULE-21-10-1/NoVariadicFunctionMacros.ql

```diff
@@ -0,0 +1,57 @@
+/**
+ * @id cpp/misra/no-variadic-function-macros
+ * @name RULE-21-10-1: The features of <cstdarg> shall not be used
+ * @description Using <cstdarg> features like va_list, va_arg, va_start, va_end and va_copy bypasses
+ *              compiler type checking and leads to undefined behavior when used incorrectly.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/misra/id/rule-21-10-1
+ *       scope/single-translation-unit
+ *       external/misra/enforcement/decidable
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.cpp.misra
+
+class VaListType extends Type {
+  VaListType() {
+    this.getName() = "va_list" or
+    this.(SpecifiedType).getBaseType() instanceof VaListType or
+    this.(TypedefType).getBaseType() instanceof VaListType
+  }
+}
+
+from Element element, string message
+where
+  not isExcluded(element, BannedAPIsPackage::noVariadicFunctionMacrosQuery()) and
+  (
+    element.(Variable).getType() instanceof VaListType and
+    (
+      if element instanceof Parameter
+      then
+        message =
+          "Declaration of parameter '" + element.(Parameter).getName() + "' of type 'va_list'."
+      else
+        message =
+          "Declaration of variable '" + element.(Variable).getName() + "' of type 'va_list'."
+    )
+    or
+    element instanceof BuiltInVarArgsStart and
+    message = "Call to 'va_start'."
+    or
+    element instanceof BuiltInVarArgsEnd and
+    message = "Call to 'va_end'."
+    or
+    element instanceof BuiltInVarArg and
+    message = "Call to 'va_arg'."
+    or
+    element instanceof BuiltInVarArgCopy and
+    message = "Call to 'va_copy'."
+    or
+    element.(TypedefType).getBaseType() instanceof VaListType and
```

Comment:
Not sure if this is worth handling, but type uses are more complex than this.

For instance, a variable or typedef could have type `va_list*`, which would not be detected. There are other missing cases technically (return types, casts, struct literals, function types).

I don't think in this case its of any serious concern, as the va_list cannot be initialized without also triggering a warning for the va_start / _arg calls etc. In theory, if we did want to handle it, I think a `TypeUse` class would make a lot of sense and _may_ be beneficial elsewhere.

Reply:
I've modified the query to use the "type uses" library. Note: most uses of `va_list` are generally prohibited - for example, as a return type or a cast. However, there are certainly cases we weren't covering - such as a variable of pointer to `va_list` type - which we're now capturing.